Java basic

miekof opened the pull request:
Created based on basic scenario.

miekof committed:
Initial commit of Java implementation

miekof committed:
Update gitignore file

Include IntelliJ specific files/directories

miekof committed:
Terminate Webdriver

After each test case, terminate webdriver. This prevents having Web
browser stays open.

eabedrapo commented on java/pom.xml:
```diff
@@ -0,0 +1,28 @@
+        </dependency>
+    </dependencies>
+
+</project>
```
You should always have an empty line at the end of your files ;)

miekof replied:
Added.

eabedrapo commented on java/src/main/java/HelloWorldTests.java:
```diff
@@ -0,0 +1,79 @@
+
+        //Find answer field
+        final String ANSWER_INPUT_XPATH = "//input[contains(@class, 'input et_pb_contact_captcha')]";
+        WebElement answerInputElement = driver.findElement(By.xpath(ANSWER_INPUT_XPATH));
```
Why did you use By.xpath here instead of using By.className? Just to demonstrate the use of XPath?

miekof replied:
Your comments brought a good question.Though best practices is use xpath as the last resort, but also this is a way to demonstrate how you can access WebElement. Hmm What do you think???

miekof replied:
changed to CSS

eabedrapo commented on java/src/main/java/HelloWorldTests.java:
```diff
@@ -0,0 +1,79 @@
+        answerInputElement.sendKeys(String.valueOf(total));
+
+        //Find the submit button, and click it
+        final String RIGHT_SUBMIT_BUTTON_XPATH = "//div[@id='et_pb_contact_form_1']//button";
```
I see that button has some classes and attributes in it. Why don't you use By.css instead of By.xpath?
(just so you know, I only use Xpath when I don't have any other alternative or if by doing so I'm making my tests more readable, which I don't think is the case here).

miekof replied:
I didn't look to use CSS, it's a habit. Let me check if I can use CSS.

eabedrapo commented on java/src/main/java/HelloWorldTests.java:
```diff
@@ -0,0 +1,79 @@
+
+        final String MESSAGE_TEXT_XPATH = "//div[@id='et_pb_contact_form_1']//div[@class='et-pb-contact-message']";
+        //Initialize wait
+        WebDriverWait wait = new WebDriverWait(driver, 10);
```
We usually prevent the use of "magic numbers" and our checkstyle script forces us to create constants to use instead of them, so in this example we usually use TEN_SECONDS instead of just 10. It won't make any difference, but tests get better documented that way 😉

miekof replied:
A good point. I will change it to constant.

eabedrapo commented on java/src/main/java/HelloWorldTests.java:
```diff
@@ -0,0 +1,79 @@
+        String successText = wait.until(ExpectedConditions.visibilityOfElementLocated(By.xpath(MESSAGE_TEXT_XPATH))).getText();
+
+        //verify obtained text is "Successful"
+        Assert.assertEquals(successText, "Success", "successful submission message should be Success");
```
In my opinion, the error message would be clearer if you just say "Fail: message should say 'Success'". But this is just an exercise, so you don't need to change it 😛

miekof replied:
I usually let the reporting system take care that portion. If I am reading the assert message, I know it's already failed.

eabedrapo commented on java/src/main/java/HelloWorldTests.java:
```diff
@@ -0,0 +1,79 @@
+        //wait for visibility of the element contains "Successful" message, then get the text.
+        String successText = wait.until(ExpectedConditions.visibilityOfElementLocated(By.xpath(MESSAGE_TEXT_XPATH))).getText();
+
+        //verify obtained text is "Successful"
```
Isn't the obtained text "Success"?

miekof replied:
A good catch!

eabedrapo commented on java/pom.xml:
```diff
@@ -0,0 +1,28 @@
+    <version>1.0-SNAPSHOT</version>
+    <properties>
+        <maven.compiler.target>1.8</maven.compiler.target>
+        <maven.compiler.source>1.8</maven.compiler.source>
```
I usually add Selenium and Runner (JUnit/TestNG) versions in this block as well so it's easier to update to a newer version in the future or having the version modified from the command line.

miekof replied:
Moved versions to properties section.

eabedrapo commented on java/pom.xml:
```diff
@@ -0,0 +1,28 @@
+        </dependency>
+    </dependencies>
+
+</project>
```
You need an empty line here too.

miekof replied:
Added

eabedrapo commented on java/pom.xml:
```diff
@@ -0,0 +1,28 @@
+            <version>3.141.59</version>
+        </dependency>
+    </dependencies>
+
```
Where are you using the compiler version that you set up in the properties section at the top?

In our pom we use that in the plugins section but I don't see a plugins section in this pom. Do you set that in another place within your project?

miekof replied:
compiler version is taken cared by
```
<maven.compiler.target>1.8</maven.compiler.target>
<maven.compiler.source>1.8</maven.compiler.source>
```
I didn't use any plugins. That's why. But if I want to run the test from command line, I need a plugin. I will update to run it via command line.

miekof replied:
Add surefire plugin--testNG can be executed from command line.
No other plugin is required at this point.

miekof committed:
Add testng file

miekof committed:
Update based on code review

* Change element locator strategy from XPath to CSS, generally CSS
selectors are faster than XPath.
* Moved ID, CLASS, CSS strings out of the test method to class level.
There is also 2nd challenge can use these values.
* Rename constants to be more clear.
* Use constants for implicit wait time to increase readability
* Fix Successful -> Success

miekof commented:
Updated based on review above.

miekof committed:
Add README.md

Explains how to run via mvn.

eabedrapo reviewed (approved):
LGTM!
